src/neo/Ledger/Blockchain.cs

```diff
@@ -62,7 +63,7 @@ private class UnverifiedBlocksList { public LinkedList<Block> Blocks = new Linke
         private readonly IActorRef txrouter;
         private readonly List<UInt256> header_index = new List<UInt256>();
         private uint stored_header_count = 0;
-        private readonly Dictionary<UInt256, Block> block_cache = new Dictionary<UInt256, Block>();
+        private readonly ConcurrentDictionary<UInt256, Block> block_cache = new ConcurrentDictionary<UInt256, Block>();
         private readonly Dictionary<uint, UnverifiedBlocksList> block_cache_unverified = new Dictionary<uint, UnverifiedBlocksList>();
```

Comment:
why not with `block_cache_unverified`?

Reply:
block_cache_unverified is not called by different threads

Reply:
It's the performance of `ConcurrentDictionary` to much worse? have you test it? Maybe we should use always a thread-safe version, otherwise it's a little dangerous.

Reply:
I searched performance comparing between these 2 & it seems that their performance are much different:
https://github.com/dotnet/runtime/issues/667
https://cc.davelozinski.com/c-sharp/dictionary-vs-concurrentdictionary
But yes we can also change block_cache_unverified to ConcurrentDictionary as there are not so many block messages on network.

Reply:
What do you think @erikzhang ?

Reply:
How about use `IImmutableDictionary`?

Reply:
If it is not used in multiple threads, it should use `Dictionary`.